Update and rename src/cal/atmospheric_correction_update4.py to src/ca…

Brian-Szutu opened the pull request:
…l/test/atmos_corr_raw_to_rad

Brian-Szutu committed:
Update and rename src/cal/atmospheric_correction_update4.py to src/cal/test/atmos_corr_raw_to_rad

Brian-Szutu committed:
Rename atmos_corr_raw_to_rad to atmos_corr_raw_to_rad.py

iparask commented:
Hello @Brian-Szutu , thank you for the Pull Request. 

Do you mind explaining what this Pull Request introduces? Also, you can assign someone specific to review this Pull Request.

I am happy to do it, but I need some extra information.

Thank you,
Giannis

bspitzbart commented:
Thanks Brian. I know this is mostly a test of branching and pull requests etc., so I don't think it needs to be merged into devel just yet.
This script nicely lays out all the steps from raw image to atmospheric correction in units of radiance.  However, to parallelize in a pipeline we will need each stage in its own kernel, ie. separate scripts for conversion to radiance, selection of ROI, spectra extraction, and atmospheric correction.

What tests have you done to show the expected result?  This will also make it easier to create and run unit tests on the individual stages.

Also, there is no need for the test folder.  It should just go in cal.  We can always backout any erroneous changes.
Finally, "THIS SCRIPT SHOULD BE PLACED IN THE DIRECTORY ABOVE ALL OF THE IMAGES TO BE PROCESSED" is undesirable. Can there be an input argument for the user to specify a path to the data dir?  Mostly likely the data and code will be separate, perhaps even on different file systems or machines.

Brian-Szutu commented:
@bspitzbart 
The only changes to my portion of this script is where the atmospheric correction text files are outputted. I did run that part with multiple files and am sure that part works. I am also moderately confident that it is able to run through multiple raw images with your raw to rad script.
The script has not fully processed raw images yet. However, if I manually end the program in the middle of processing, a 5 million byte file is outputted. This more or less matches the size of the output file of your script.
I believe I can have the script take in user input for the directory where the images are located in. It may be until the weekend when I can work on that though.

Brian-Szutu committed:
Update atmos_corr_raw_to_rad.py

Brian-Szutu committed:
Update atmos_corr_raw_to_rad.py

Brian-Szutu committed:
Rename src/cal/test/atmos_corr_raw_to_rad.py to src/cal/testatmos_corr_raw_to_rad.py

Brian-Szutu committed:
Rename testatmos_corr_raw_to_rad.py to test_atmos_corr_raw_to_rad.py

Brian-Szutu committed:
Update test_atmos_corr_raw_to_rad.py

Brian-Szutu committed:
Update test_atmos_corr_raw_to_rad.py

Brian-Szutu committed:
Update test_atmos_corr_raw_to_rad.py

Brian-Szutu committed:
Create rad_to_refl.py

iparask commented:
@Brian-Szutu feel free to ping @bspitzbart and I when you want us to review this pull request

Great work! keep up!

bspitzbart commented:
date_distance in rad_to_refl.py should be moved to a static library file.  What is the source of date_distance? This is one unit test to check the values are correct with an independent source.
(I'm looking for a table or function that we don't need to maintain to return earth-sun distance)

Also, lines 191-195 in rad_to_refl.py are not needed.
collect band metadata
```python
abscalfactor = float(rt.find('ABSCALFACTOR').text)
effbandwidth = float(rt.find('EFFECTIVEBANDWIDTH').text)
print(bands[i])
print(src.read(i+1)[0,0]," ",abscalfactor," ",effbandwidth)
```

helen-eifert commented:
@bspitzbart , the table the date_distance values came from was given by Mark in his image processing directions. The file exists in the Google Drive under landcover, titled "ENVI_AntarcticData_Instructions". I believe Brian did quite a bit of searching for a document with all these values and couldn't find one in an appropriate format. Perhaps we copy it over into a separate document to reference for the future? 

@Brian-Szutu , did you make the following changes to the script here in Git-removing lines 191-195 as Brad stated above and adjusting the band math equation that we covered in our meeting yesterday?

Finally, as for renaming...I'd like to request this script simply be titled "refl" and for the "raw_to_rad" script to simply be "rad"...because the reflectance script can take both a radians input for top of atmosphere correction and atmcorr for atmospherically corrected, I don't want to specify the input in the name. Accordingly, the other scripts should have a similar format. As for the atmospheric correction steps, if we can have the atmospheric correction regression tests script (the first script I had you develop, Brian) separate and called "atmcorr_regressions" and then the spectral math will be "atmcorr_removal" and each of the other steps can be named as we develop them how we see fit with atmcorr (e.g. atmcorr_spec_collection, atmcorr_ROI_finder, etc). Let me know if there is a better place I can formally put in this request and if it is worth creating some kind of visual graphic resource.

iparask commented on src/cal/rad_to_refl.py:
```diff
@@ -0,0 +1,201 @@
+
+# This is a large dictionary of the Earth-Sun distance in AU.
+# Goes by date. Hopefully nothing was misinputted
+date_distance = {'JAN01':0.98331,'JAN02':0.98330,'JAN03':0.98330,'JAN04':0.98330,
```
May I suggest that this table is moved to a separate file? It can be imported in a way like:
```python
from distances import date_distance
```
It will also help with the readability of this file.

iparask commented on src/cal/rad_to_refl.py:
```diff
@@ -0,0 +1,201 @@
+                 'DEC29':0.98337,'DEC30':0.98335,'DEC31':0.98333,
+                 }
+
+#///////////
```
This looks more like a script, rather than a program or a function. I would suggest that the following become part of a function which is then called from a main.

```python

def YourFunctionName():
    ...

if __name__=='__main__':
    YourFunctionName()
```

iparask commented on src/cal/test_atmos_corr_raw_to_rad.py:
```diff
@@ -2,21 +2,36 @@
 Programmer: Brian Szutu
```
Well done for the banner! It really helps. Do you mind changing it in the following way:

```
"""
Author: Your Name
email: your email (optional)
License: MIT
Copyright: 2018-2019
"""
```

iparask commented on src/cal/test_atmos_corr_raw_to_rad.py:
```diff
@@ -64,14 +79,16 @@ def reader(file):
 
 
-def writer(file, file_name, pass_fail_stat_arr, pass_fail_arr, intercept_arr, set_check):
+def writer(file, file_name, output_dir, pass_fail_stat_arr, \
```
`\` is not really needed. You can remove it.

iparask commented on src/cal/test_atmos_corr_raw_to_rad.py:
```diff
@@ -84,7 +101,7 @@ def writer(file, file_name, pass_fail_stat_arr, pass_fail_arr, intercept_arr, se
 
     # Creates a file to APPEND text to the end of it
-    file_write = open('Output Files\\'+file_name+'.txt', 'a+')
+    file_write = open(os.path.join(output_dir, file_name+'.txt'), 'a+')
```
Can you add spaces between `+`?

iparask commented on src/cal/test_atmos_corr_raw_to_rad.py:
```diff
@@ -278,34 +299,35 @@ def avg_intercept(total_intercept_arr, file_count):
-    band5_avg = band5_sum / file_count
-    band6_avg = band6_sum / file_count
-    band7_avg = band7_sum / file_count
+    band1_avg = band1_sum / txt_count
```
Wouldn't a numpy array make more sense? Or there is a reason these values are separated?

iparask commented on src/cal/test_atmos_corr_raw_to_rad.py:
```diff
@@ -278,34 +299,35 @@ def avg_intercept(total_intercept_arr, file_count):
     None
     """
-    file_write = open('Output Files\\'+file_name+'.txt', 'a+')
+    file_write = open(os.path.join(output_dir, file_name+'.txt'), 'a+')
```
Can you add spaces around `+`

iparask commented on src/cal/test_atmos_corr_raw_to_rad.py:
```diff
@@ -325,21 +347,25 @@ def main():
     """
+
+    # Prints the directory the script is in
+    print("Current directory: " + os.path.dirname(os.path.abspath(__file__)))
```
This is a debug message, right?

iparask commented on src/cal/test_atmos_corr_raw_to_rad.py:
```diff
@@ -325,21 +347,25 @@ def main():
-    # If the folder doesn't exist, then it will be made.
-    if not os.path.exists(final_dir):
-        os.makedirs(final_dir)
+    working_dir = input("Please input the FULL path to the directory with the images: \n")
```
Instead of asking for the user to provide input as the script runs, you can add input arguments. An example can be found [here](https://github.com/iceberg-project/Seals/blob/master/src/entk_script/entk_script.py#L154). 

Eventually this script will run without any interaction by the user. As a result any input during execution time will not be possible. Providing everything during the python call is preferable. It will also minimize the changes that may be needed for when we run it on a Supercomputer

iparask commented on src/cal/test_atmos_corr_raw_to_rad.py:
```diff
@@ -414,20 +460,80 @@ def main():
+            for f in tif_files:
+
+                # Sees if an output file for the raw image being analyzed exists...
+                rad_file_exists = os.path.isfile(os.path.join(output_dir, f+'_rad.tif'))
```
Can you add space around `+`

iparask reviewed (commented):
This is nice work and very well done! I really like that the code is significantly commented. It will make all our lives easier later in the project.

I've left some stylistic and usability comments, before going in the PR in more detail. I trust that you know your science and algorithms better than me, so my contribution will mainly be in code quality :)

A general suggestion, it is better not to have prints for information that is communicated by a simple check in folders and such. Instead your prints can show things that trigger specific parts of the code and what triggered them. That way it will be easier to verify the correctness of the execution (if it makes sense). [Here](https://github.com/iceberg-project/Seals/blob/experimental/data-design/src/iceberg/seals/iceberg_queue/Queue.py#L136) is an example.

Brian-Szutu committed:
Add files via upload

Brian-Szutu committed:
Update and rename rad_to_refl.py to reflectance.py

Brian-Szutu committed:
Update test_atmos_corr_raw_to_rad.py

Brian-Szutu committed:
Update test_atmos_corr_raw_to_rad.py

helen-eifert committed:
Rename test_atmos_corr_raw_to_rad.py to rad.py

helen-eifert committed:
Rename reflectance.py to refl.py

Brian-Szutu committed:
Update refl.py

Brian-Szutu committed:
Update rad.py

rad.py originally had the atmospheric regression in it. The atmospheric regression is now in its own separate script

Brian-Szutu committed:
Update rad.py

Brian-Szutu committed:
Update rad.py

Brian-Szutu committed:
Fixes issue #14

Will now only read P1BS images and their .xml files. S1BS images and .xml files are excluded. Also fixed a minor grammar error in a print statement.

Brian-Szutu committed:
Fixes issue #15

Added input through console to allow for the directory to be passed outside of Python. Added checks to made for no .xml files and no atmospherically corrected .tif files to print relevant messages.

Commit:
Remove old, unused scripts

Brian-Szutu committed:
Fixes issue #16

Fixes how the outputted .tif file is named from nameOfFile.tif_rad.tif to nameOfFile_rad.tif

Brian-Szutu committed:
Fixes issue #16

Fixes how the .tif file is named, from nameOfFile.tif_refl.tif to nameOfFile_refl.tif.

helen-eifert committed:
parenthesis error in line 94-95

Brian-Szutu committed:
Fixes issue #17

Forgot a parenthesis in line 121. Should be fixed now assuming nothing else is wrong.

Brian-Szutu committed:
Fixes issues #18, #19, #20, #21, and #22

This update to the script does the following:
Allows the script to use rad.tif files if atmocorr.tif files are not available in order to create the refl.tif files. Also renames the atmcorr_files and atmcorr_count to refl_ready_images and refl_ready_count, respectively.
Removes some remnant lines from rad.py that are not relevant to the creation of refl.tif files
Renames the rad variable in now line 170 to refl.
Prints a message indicating which image was used to successfully create a refl.tif file.

helen-eifert committed:
Move earth_sun_dist.py to appropriate lib folder

helen-eifert committed:
Delete tmp

helen-eifert committed:
Update README.md

helen-eifert committed:
Update README.md

helen-eifert committed:
Update README.md

helen-eifert committed:
Converts 'MEANSUNEL' from str to float

Brian-Szutu committed:
Fixes issue #23

Fixed a misplaced parenthesis in line 163. Also closes the file being written to upon completion.

helen-eifert committed:
Name lookup adjustment

Line 114

helen-eifert committed:
Converts MEANSUNEL from str to float

helen-eifert committed:
Addresses Issues 24 & 25

Tries to also address Issue 26.

bspitzbart committed:
change dtype to float32

helen-eifert committed:
Re adds numpy packages

numpy wasn't being used and is now with float32 conversion

helen-eifert committed:
Changes sine value to degrees

bspitzbart commented:
@Brian-Szutu, I finally got rad.py to find my input files, but I am curious why you look down two levels in the directory tree instead of just in the directory specified by -ip?
I ran 'python rad.py -ip IN' (with the tif and xml files in IN).  It does not find anything to process. But with the input files in IN/Test, using the same command, it works. 
rad.py lines 67-81
# for each file/folder in the specified directory...
    for file in os.listdir(working_dir):
        # if there is a . or the file is named Output Files...
        if "." in file or file == 'Output Files':
            # ...don't do anything
            continue
        # else, append the file to the folders list
        else:
            folders.append(file)

    # for each folder in the folders list...
    for folder in folders:
        # Initialize a variable to save the name of the .xml file.
        # Initialize a variable to count the number of .xml files.
       xml_file = ''; xml_count = 0
.
.
.

You could remove lines 69-75 and just look for input files with:
for file in os.listdir(working_dir):
        # Initialize a variable to save the name of the .xml file.
        # Initialize a variable to count the number of .xml files.
       xml_file = ''; xml_count = 0

There may be a reason you want to keep the input in separate directories (NW region, SE region, etc. or Coastline, Plains, or whatever) but then the documentation should be clearer.  Let me know what you think.  Otherwise it seems to be working great (I ran on Linux)!

bspitzbart commented:
Ah, now I see refl.py expects the subdir to be named by date.  That is a lot of unnecessary setup on the users part.  A safer way to get the date is from the xml file. There is a TLCTIME tag.

bspitzbart commented:
Also, since you moved earth_sun_dist, you need to reset the import path.
import sys
sys.path.insert(0, "./..")
from lib.earth_sun_dist import date_distance

helen-eifert commented:
> Also, since you moved earth_sun_dist, you need to reset the import path.
> import sys
> sys.path.insert(0, "./..")
> from lib.earth_sun_dist import date_distance

@bspitzbart what should I input for the path? On my system, the GitHub tree is located in my user documents folder, but that won't be repeatable on other devices. Can I just reference the lib folder since it is still otherwise within GitHub/LandCover/src?

bspitzbart commented:
sys.path.insert(0, "./..") says look in the current folder and up one folder.  So you can just cut and paste that 3 lines of code to reset the path.

helen-eifert committed:
Fixes Issue #28 and adjusts for new file structure

Fixes the reflectance equation to convert meansunel degree values to radians to work with math.sin function. No more inversion!
Also fixes the lookup for earth_sun_dist to find in the appropriate lib folder instead of cal

Commit:
Merge branch 'devel' into Brian-Szutu-patch-1
to resolve conflict caused by deleted file mk_rad.py
